File: LFUCache.cpp
#include<iostream>
#include<unordered_map>
#include<unordered_set>
#include<set>

using namespace std;

class LFUCache {
private:
	struct Node {
		int key;
		int value;
		int times;
		Node *pre;
		Node *next;
		Node(int key, int value) : key(key), value(value), times(1), pre(NULL), next(NULL) {}
	};
private:
	int capacity;
	unordered_map<int, Node*> key_value;
	set<int> times;
	unordered_map<int, pair<Node*, Node*> > times_map;
	
public:
    LFUCache(int capacity) {
        this->capacity = capacity;
    }
    
    int get(int key) {
        if(key_value.find(key) != key_value.end()) {
        	
        	Node *cur = key_value[key];
        	delete_one(cur);
//        	Node *begin = times_map[cur->times].first;
//        	Node *tail = times_map[cur->times].second;
//        	
//        	if(begin == tail) {
//        		times_map.erase(cur->times);
//        		times.erase(cur->times);
//			}else {
//				if(cur == begin) {
//					begin = begin->next;
//					begin->pre = NULL;
//					times_map[cur->times].first = begin;
//				}else if(cur == tail) {
//					tail = tail->pre;
//					tail->next = NULL;
//					times_map[cur->times].second = tail;
//				}else {
//					cur->pre->next = cur->next;
//					cur->next->pre = cur->pre;
//				}
//			}
        	
        	cur->times++;
//        	if(times_map.find(cur->times) == times_map.end()) {
//        		times_map[cur->times] = make_pair(cur, cur);
//        		times.insert(cur->times);
//			}else {
//				times_map[cur->times].second->next = cur;
//				cur->pre = times_map[cur->times].second;
//				cur->next = NULL;
//				times_map[cur->times].second = cur;
//			}
			add_one(cur);
        	
        	return cur->value;
        	
		}else return -1;
    }
    
    void delete_one(Node *cur) {
		Node *begin = times_map[cur->times].first;
    	Node *tail = times_map[cur->times].second;
    	cout<<"key: "<<begin->key<<endl;
    	cout<<"key2: "<<tail->key<<endl;
    	
    	if(begin == tail) {
    		times_map.erase(cur->times);
    		times.erase(cur->times);
		}else {
			if(cur == begin) {
				cout<<"---1---"<<endl;
				begin = begin->next;
				begin->pre = NULL;
				times_map[cur->times].first = begin;
			}else if(cur == tail) {
				tail = tail->pre;
				tail->next = NULL;
				times_map[cur->times].second = tail;
			}else {
				cur->pre->next = cur->next;
				cur->next->pre = cur->pre;
			}
		}
	}
	
	void add_one(Node *cur) {
		if(times_map.find(cur->times) == times_map.end()) {
    		times_map[cur->times] = make_pair(cur, cur);
    		times.insert(cur->times);
		}else {
			times_map[cur->times].second->next = cur;
			cur->pre = times_map[cur->times].second;
			cur->next = NULL;
			times_map[cur->times].second = cur;
		}
	}
    
    void put(int key, int value) {
        if(key_value.find(key) != key_value.end()) {
        	Node *cur = key_value[key];
        	delete_one(cur);
        	cur->value = value;
        	cur->times++;
        	add_one(cur);
		}else if(capacity > 0) {
			cout<<"capacity: "<<capacity<<endl;
			Node *new_one = new Node(key, value);
			add_one(new_one);
			capacity--;
			key_value[key] = new_one;
		}else {
			if(key_value.empty()) return;
			int time_to_delete = *(times.begin());
			cout<<"begin: "<<time_to_delete<<endl;
			Node *to_delete = times_map[time_to_delete].first;
			key_value.erase(to_delete->key);
			delete_one(to_delete);
			
			Node *new_one = new Node(key, value);
			add_one(new_one);
			key_value[key] = new_one;
		}
    }
};

int main() {
	
	LFUCache *cache = new LFUCache(2);
	cache->put(1, 1);
	cache->put(2, 2);
	cout<<cache->get(1)<<endl;
	cache->put(3, 3);
	cout<<cache->get(2)<<endl;
	cout<<cache->get(3)<<endl;
	cache->put(4, 4);
	cout<<cache->get(1)<<endl;
	cout<<cache->get(3)<<endl;
	cout<<cache->get(4)<<endl;
	
	return 0;
	
}

File: LFUCache.py
class LFUCache:
    class Node:
        def __init__(self, key, value):
            self.key = key
            self.value = value
            self.count = 1
            self.next = None
            self.pre = None

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.minFreq = 0
        self.m = {}
        self.freq = {}

    def remove(self, cur):
        head, tail = self.freq[cur.count][0], self.freq[cur.count][1]
        if head == tail:
            del self.freq[cur.count]
            if self.minFreq == cur.count:
                self.minFreq += 1
        elif cur == head:
            head = head.next
            head.pre = None
            self.freq[cur.count][0] = head
        elif cur == tail:
            tail = tail.pre
            tail.next = None
            self.freq[cur.count][1] = tail
        else:
            cur.pre.next = cur.next
            cur.next.pre = cur.pre

    def add(self, cur):
        if cur.count in self.freq:
            self.freq[cur.count][1].next = cur
            cur.pre = self.freq[cur.count][1]
            cur.next = None
            self.freq[cur.count][1] = cur
        else:
            self.freq[cur.count] = [cur, cur]

    def get(self, key: int) -> int:
        if key not in self.m:
            return -1
        self.remove(self.m[key])
        self.m[key].count += 1
        self.add(self.m[key])
        return self.m[key].value

    def put(self, key: int, value: int) -> None:
        if self.get(key) != -1:
            self.m[key].value = value
            return
        elif self.capacity > 0:
            self.capacity -= 1
        else:
            if not self.m:
                return
            cur = self.freq[self.minFreq][0]
            del self.m[cur.key]
            self.remove(cur)
        self.m[key] = self.Node(key, value)
        self.add(self.m[key])
        self.minFreq = 1


if __name__ == '__main__':
    cache = LFUCache(2)
    cache.put(1, 1)
    cache.put(2, 2)
    print(cache.get(1))
    cache.put(3, 3)
    print(cache.get(2))
    print(cache.get(3))
    cache.put(4, 4)
    print(cache.get(1))
    print(cache.get(3))
    print(cache.get(4))
